Fixes #6342 - merge compute profile when attributes in api host#create

tstrachota opened the pull request.

Commit:
Fixes #6342 - merge compute profile when attributes in api host#create

API v2 host#craete call now merges user specified compute attributes
with compute profile. UI functionality remains untouched.

domcleal commented:
#1551 did something similar, but we agreed that it seemed like the compute profile should be merged with the compute_attributes rather than ignoring empty hashes, do you agree?

tstrachota commented:
Merging makes sense. I didn't know about the previous discussion. I'll take a look at it.

tstrachota commented:
@domcleal updated. API now merges the parameters while UI remains untouched.

tstrachota commented:
[test]

domcleal commented:
[test] because of 85e82d0 being merged.

domcleal commented:
Would you mind rebasing this before review?  Seems there's a subtle conflict with the other PR causing a test failure.

tstrachota commented:
Rebased, fixing the test failures now...

tstrachota commented:
Rebased, tests fixed, ready for review.
I had to heighten the class length cop because of the host managed class.

tstrachota commented:
Rebased

tstrachota commented:
[test]

domcleal commented on app/services/compute_attribute_merge.rb:
```diff
@@ -0,0 +1,12 @@
+    host.compute_attributes ||= {}
+
+    vm_attrs = compute_attrs.vm_attrs
+    vm_attrs.deep_merge!(host.compute_attributes)
```
Nitpick, you might want to avoid using `!` to not change the compute profile itself in memory, just run deep_merge and assign the result.

domcleal commented on app/services/compute_attribute_merge.rb:
```diff
@@ -0,0 +1,12 @@
+  def run(host, compute_attrs)
+    return if compute_attrs.nil?
+
+    host.compute_attributes ||= {}
```
Should this be with indifferent access?

domcleal commented on test/factories/host_related.rb:
```diff
@@ -98,6 +98,19 @@ def set_nic_attributes(host, attributes, evaluator)
     sequence(:name) { |n| "hal900#{n}" }
   end
 
+  factory :compute_attribute do
```
Nitpick, these might be better in the compute_resources.rb file.

domcleal commented on test/factories/host_related.rb:
```diff
@@ -167,9 +180,20 @@ def set_nic_attributes(host, attributes, evaluator)
       uuid Foreman.uuid
       association :compute_resource, :factory => :ec2_cr
-      after(:build) { |host| host.expects(:queue_compute) }
+      after(:build) { |host, evaluator| host.expects(:queue_compute) if evaluator.expect_queue_compute }
```
Or perhaps change this to `before(:create)` so it's only expected if the host is going to be saved, and won't be expected when it's built only.

domcleal commented on test/factories/host_related.rb:
```diff
@@ -167,9 +180,20 @@ def set_nic_attributes(host, attributes, evaluator)
+    trait :with_compute_profile do
+      after(:build) do |host|
+        host.compute_profile = FactoryGirl.create(:compute_profile)
+        host.compute_profile.compute_attributes << FactoryGirl.build(:compute_attribute, :compute_resource => host.compute_resource)
```
Should this be in the :compute_profile create call above?  e.g. a trait of the profile?

domcleal commented:
Sorry for the delay in re-reviewing this PR!

It looks good, and the re-use of the interfaces is nice.  The only thing I see missing is that there's no functional test coverage to check the attrs merger is called from the hosts API.  It doesn't have to be a deep test as it's covered well elsewhere, but an expectation at least would be good.

tstrachota commented:
@domcleal I fixed the comments and added simple functional tests to check application of the merge classes. Nothing nice (`any_instance` :-/ ) but it serves the purpose.

domcleal commented:
Merged as 4799063, thanks @tstrachota!
